azure-servicebus/azure/servicebus/aio/async_base_handler.py

```diff
@@ -0,0 +1,169 @@
+        if not self.running:
+            raise InvalidHandlerState("Client connection is closed.")
+
+        mgmt_msg = Message(
+            body=message,
+            properties=MessageProperties(
+                reply_to=self.mgmt_target,
+                encoding=self.encoding,
+                **kwargs))
+        try:
+            return await self._handler.mgmt_request_async(
+                mgmt_msg,
+                operation,
+                op_type=b"entity-mgmt",
+                node=self.mgmt_target.encode(self.encoding),
+                timeout=5000,
+                callback=callback)
+        except Exception as exp:  # pylint: disable=broad-except
+            raise ServiceBusError("Management request failed: {}".format(exp), exp)
+
+    async def _handle_exception(self, exception):
+        if isinstance(exception, (errors.LinkDetach, errors.ConnectionClose)):
+            if exception.action and exception.action.retry and self.auto_reconnect:
+                log.info("Async handler detached. Attempting reconnect.")
+                await self.reconnect()
+            elif exception.condition == constants.ErrorCodes.UnauthorizedAccess:
+                log.info("Async handler detached. Shutting down.")
+                error = ServiceBusAuthorizationError(str(exception), exception)
+                await self.close(exception=error)
+                raise error
+            else:
+                log.info("Async handler detached. Shutting down.")
+                error = ServiceBusConnectionError(str(exception), exception)
+                await self.close(exception=error)
+                raise error
+        elif isinstance(exception, errors.MessageHandlerError):
+            if self.auto_reconnect:
+                log.info("Async handler error. Attempting reconnect.")
+                await self.reconnect()
+            else:
+                log.info("Async handler error. Shutting down.")
+                error = ServiceBusConnectionError(str(exception), exception)
+                await self.close(exception=error)
+                raise error
+        elif isinstance(exception, errors.AMQPConnectionError):
+            message = "Failed to open handler: {}".format(exception)
+            raise ServiceBusConnectionError(message, exception)
+        else:
+            log.info("Unexpected error occurred (%r). Shutting down.", exception)
+            error = ServiceBusError("Handler failed: {}".format(exception), exception)
+            await self.close(exception=error)
+            raise error
+
+    async def reconnect(self):
+        """If the handler was disconnected from the service with
+        a retryable error - attempt to reconnect."""
+        await self._handler.close()
+        self._build_handler()
+        await self.open()
+
+    async def open(self):
+        """Open handler connection and authenticate session."""
+        if self.running:
+            return
+        self.running = True
```

Comment:
My thread-safe spider-sense is screaming here. Thoughts?

Reply:
Yes I agree - though I think I'm inclined to simply document the object as not thread-safe. At least for preview. As I think if you tried to run a single Receiver/Sender with multiple threads it may encounter other issues besides this one. 

For preview I'm tempted to make my recommendation:
"If you wish to run multi-threaded senders or receivers, we recommend using one sender/receiver per thread"

Then I can revisit it later with more testing. Thoughts?